File: APCS202010_人口遷移.cpp
#include <bits/stdc++.h>

using namespace std;

#define N 60
typedef long long ll;

int mat[N][N], refer[N][N];

int main()
{
    // freopen("P_6_19_3.in", "r", stdin);
    int r, c, k, m;
    cin >> r >> c >> k >> m;
    memset(mat, -1, sizeof(mat));
    memset(refer, -1, sizeof(refer));
    for (int i = 1; i <= r; ++i)
        for (int j = 1; j <= c; ++j)
            cin >> mat[i][j];

    int di[4] = {-1, 0, 1, 0}, dj[4] = {0, 1, 0, -1};
    while (m--)
    {
        for (int i = 1; i <= r; ++i)
        {
            for (int j = 1; j <= c; ++j)
            {
                if (mat[i][j] >= 0)
                {
                    refer[i][j] = mat[i][j];
                    for (int d = 0; d < 4; ++d)
                    {
                        int ni = i + di[d], nj = j + dj[d];
                        if (mat[ni][nj] >= 0)
                        {
                            refer[i][j] += mat[ni][nj] / k;
                            refer[i][j] -= mat[i][j] / k;
                        }
                    }
                }
            }
        }
        for (int i = 1; i <= r; ++i)
            swap(mat[i], refer[i]);
    }
    int mx = 0, mn = 1e9;
    for (int i = 1; i <= r; ++i)
    {
        for (int j = 1; j <= c; ++j)
        {
            if (mat[i][j] != -1)
            {
                mx = max(mx, mat[i][j]);
                mn = min(mn, mat[i][j]);
            }
        }
    }
    cout << mn << '\n'
         << mx;
    return 0;
}

File: APCS202010_人口遷移.py
r, c, k, m = map(int, input().split())
r += 2
c += 2
mat = [[] for _ in range(r)]
mat[0] = mat[r-1] = [-1 for _ in range(c)]
temp = [[-1]*c for _ in range(r)]
for i in range(1, r-1):
    mat[i] = [-1]+[int(x) for x in input().split()]+[-1]

di = [-1, 0, 1, 0]
dj = [0, 1, 0, -1]
for day in range(m):
    for i in range(1, r-1):
        for j in range(1, c-1):
            temp[i][j] = mat[i][j]
            if mat[i][j] >= 0:
                for dir in range(4):
                    ni, nj = i+di[dir], j+dj[dir]
                    if mat[ni][nj] >= 0:
                        temp[i][j] -= mat[i][j]//k
                        temp[i][j] += mat[ni][nj]//k
    mat, temp = temp, mat

print(min(mat[i][j] for i in range(r) for j in range(c) if mat[i][j] != -1))
print(max(mat[i][j] for i in range(r) for j in range(c) if mat[i][j] != -1))
